// fs.realpath/index.js
export default realpath
realpath.realpath = realpath
realpath.sync = realpathSync
realpath.realpathSync = realpathSync
realpath.monkeypatch = monkeypatch
realpath.unmonkeypatch = unmonkeypatch

import { realpath as _realpath, realpathSync as _realpathSync } from 'fs'
var origRealpath = _realpath
var origRealpathSync = _realpathSync

var version = process.version
var ok = /^v[0-5]\./.test(version)
import { realpath as __realpath, realpathSync as __realpathSync } from './old.js'

function newError (er) {
  return er && er.syscall === 'realpath' && (
    er.code === 'ELOOP' ||
    er.code === 'ENOMEM' ||
    er.code === 'ENAMETOOLONG'
  )
}

function realpath (p, cache, cb) {
  if (ok) {
    return origRealpath(p, cache, cb)
  }

  if (typeof cache === 'function') {
    cb = cache
    cache = null
  }
  origRealpath(p, cache, function (er, result) {
    if (newError(er)) {
      __realpath(p, cache, cb)
    } else {
      cb(er, result)
    }
  })
}

function realpathSync (p, cache) {
  if (ok) {
    return origRealpathSync(p, cache)
  }

  try {
    return origRealpathSync(p, cache)
  } catch (er) {
    if (newError(er)) {
      return __realpathSync(p, cache)
    } else {
      throw er
    }
  }
}

function monkeypatch () {
  _realpath = realpath
  _realpathSync = realpathSync
}

function unmonkeypatch () {
  _realpath = origRealpath
  _realpathSync = origRealpathSync
}
